src/miniscript/iter.rs

```diff
@@ -0,0 +1,703 @@
+    /// Creates a new [Iter] iterator that will iterate over all [Miniscript] items within
+    /// AST by traversing its branches. For the specific algorithm please see
+    /// [Iter::next] function.
+    pub fn iter(&self) -> Iter<PK> {
+        Iter::new(self)
+    }
+
+    /// Creates a new [PubkeyIter] iterator that will iterate over all plain public keys (and not
+    /// key hash values) present in [Miniscript] items within AST by traversing all its branches.
+    /// For the specific algorithm please see [PubkeyIter::next] function.
+    pub fn iter_pubkeys(&self) -> PubkeyIter<PK> {
+        PubkeyIter::new(self)
+    }
+
+    /// Creates a new [PubkeyHashIter] iterator that will iterate over all plain public keys (and not
+    /// key hash values) present in Miniscript items within AST by traversing all its branches.
+    /// For the specific algorithm please see [PubkeyHashIter::next] function.
+    pub fn iter_pubkey_hashes(&self) -> PubkeyHashIter<PK> {
+        PubkeyHashIter::new(self)
+    }
+
+    /// Creates a new [PubkeyAndHashIter] iterator that will iterate over all plain public keys and
+    /// key hash values present in Miniscript items within AST by traversing all its branches.
+    /// For the specific algorithm please see [PubeyAndHashIter::next] function.
+    pub fn iter_pubkeys_and_hashes(&self) -> PubkeyAndHashIter<PK> {
+        PubkeyAndHashIter::new(self)
+    }
+
+    /// Enumerates all child nodes of the current AST node (`self`) and returns a `Vec` referencing
+    /// them.
+    pub fn branches(&self) -> Vec<&Miniscript<PK>> {
+        use Terminal::*;
+        match &self.node {
+            &PkK(_) | &PkH(_) | &Multi(_, _) => vec![],
+
+            &Alt(ref node)
+            | &Swap(ref node)
+            | &Check(ref node)
+            | &DupIf(ref node)
+            | &Verify(ref node)
+            | &NonZero(ref node)
+            | &ZeroNotEqual(ref node) => vec![node.deref()],
+
+            &AndV(ref node1, ref node2)
+            | &AndB(ref node1, ref node2)
+            | &OrB(ref node1, ref node2)
+            | &OrD(ref node1, ref node2)
+            | &OrC(ref node1, ref node2)
+            | &OrI(ref node1, ref node2) => vec![node1.deref(), node2.deref()],
+
+            &AndOr(ref node1, ref node2, ref node3) => {
+                vec![node1.deref(), node2.deref(), node3.deref()]
+            }
+
+            &Thresh(_, ref node_vec) => node_vec.iter().map(Arc::deref).collect(),
+
+            _ => vec![],
+        }
+    }
+
+    /// Returns `Vec` with cloned version of all public keys from the current miniscript item,
+    /// if any. Otherwise returns an empty `Vec`.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    /// To obtain a list of all public keys within AST use [`iter_pubkeys()`] function, for example
+    /// `miniscript.iter_pubkeys().collect()`.
+    pub fn get_pubkeys(&self) -> Vec<PK> {
```

Comment:
Is it necessary for these functions to be `pub`? Seems like this is only required for internal iterator purposes. The name might even be confusing for users(who did not look at the doc comment) as they might expect a recursive implementation. 
I think we can get rid of the pub of this and other similar functions.

Reply:
I am using these function in my downstream dependency on this library. And in general, it is quite useful to be able to get a list of all unique public keys and/or hashes from a given script of any complexity

Reply:
I think we should either 
1. change the name of the function to highlight that it is not considering the descendants. 
OR
2. Make the `is_leaf()` type function and downstream call it `miniscript.is_leaf() && miniscript.iter_pubkeys.collect()`